src/game_destiny.cpp

```diff
@@ -0,0 +1,350 @@
+		? static_cast<Destiny::Language>(language)
+		: Destiny::Language::ENGLISH;
+	_extra = extra;
+	_decimalComma = _language == Destiny::Language::DEUTSCH;
+
+	EvaluateExtraFlags();
+	CheckVersionInfo();
+
+	// Init containers
+	_dwords.resize(dwordSize);
+	_floats.resize(floatSize);
+	_strings.resize(stringSize);
+
+	// TODO: Init File container
+	// TODO: Init ClientSocket container
+
+	// Debug
+	Output::Debug("[Destiny] Initialized");
+	Output::Debug("[Destiny] Language: {}", _decimalComma ? "Deutsch" : "English");
+	Output::Debug("[Destiny] DLL Version: {}", _dllVersion.toString());
+	Output::Debug("[Destiny] Dwords: {}", dwordSize);
+	Output::Debug("[Destiny] Floats: {}", floatSize);
+	Output::Debug("[Destiny] Strings: {}", stringSize);
+	Output::Debug("[Destiny] RPG Maker version: {}", _rm2k3 ? 2003 : 2000);
+}
+
+void Game_Destiny::Terminate()
+{
+	_dwords.clear();
+	_floats.clear();
+	_strings.clear();
+
+	// TODO [XGB]: Clear File container
+	// TODO [XGB]: Clear ClientSocket container
+}
+
+bool Game_Destiny::Main(SaveEventExecFrame& frame)
+{
+	const char* script;
+	InterpretFlag flag;
+
+	script = _interpreter.MakeString(frame);
+	flag = InterpretFlag::IF_EXIT;
+
+	_interpreter.CleanUpData();
+	_interpreter.LoadInterpretStack();
+
+	while (! _interpreter.IsEndOfScript())
+	{
+		if (_interpreter.IsEndOfLine())
+		{
+			_interpreter.ScriptNextChar();
+		}
+
+		flag = _interpreter.Interpret();
+		if (flag == InterpretFlag::IF_EXIT)
+		{
+			break;
+		}
+	}
+
+	Output::Debug("DestinyScript Code:\n{}", script);
+	_interpreter.FreeString();
+
+	return true;
+}
+
+void Game_Destiny::EvaluateExtraFlags()
+{
+	_trueColor = (_extra & Destiny::DF_TRUECOLOR) << 9;
+	_protect = _extra & Destiny::DF_PROTECT;
+}
+
+void Game_Destiny::CheckVersionInfo()
+{
+	uint32_t gameVersionMajor;
+
+	gameVersionMajor = _gameVersion.major;
+
+	if (! (gameVersionMajor == 0x2000 || gameVersionMajor == 0x2003))
+	{
+		Output::Error("[Destiny]: {} is not a valid version", gameVersionMajor);
+	}
+
+	_rm2k3 = gameVersionMajor == 0x2003;
+}
+
+
+// =-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
+//			MainFunctions / Interpreter
+// =-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
+
+Interpreter::Interpreter()
+{
+	CleanUpData();
+	_destinyScript = nullptr;
+	_scriptPtr = nullptr;
+}
+
+const char* Interpreter::MakeString(SaveEventExecFrame& frame)
+{
+	std::string code;
+	size_t length;
+	char* destinyScript;
+
+	int32_t& current = frame.current_command;
+	const std::vector<EventCommand>& cmdList = frame.commands;
+	std::vector<EventCommand>::const_iterator it = cmdList.begin() + current++;
+
+	code = ToString((*it++).string);
+
+	while (it != cmdList.cend() && it->code == static_cast<int32_t>(EventCommand::Code::Comment_2))
+	{
+		code += '\n';
+		code += ToString((*it++).string);
+		++current;
+	}
+
+	length = code.length() + 1;
+	destinyScript = new char[length];
+	strcpy_s(destinyScript, length, code.c_str());
+
```

Comment:
``strcpy_s`` is a Microsoft extension.

In general you do not need ``new`` or ``malloc`` in modern C++ code. You could just change the type of ``_destinyScript`` to ``std::string`` and do ``destinyScript = code``.

You can still do the ``_scriptPtr`` stuff via ``_scriptPtr = &destintyScript.data()``.

Reply:
Like this?
```cpp
private:
	// Member data
	std::string _destinyScript;
```
```cpp
const char* Interpreter::MakeString(SaveEventExecFrame& frame)
{
	std::string code;

	int32_t& current = frame.current_command;
	const std::vector<EventCommand>& cmdList = frame.commands;
	std::vector<EventCommand>::const_iterator it = cmdList.begin() + current++;

	code = ToString((*it++).string);

	while (it != cmdList.cend() && it->code == static_cast<int32_t>(EventCommand::Code::Comment_2))
	{
		code += '\n';
		code += ToString((*it++).string);
		++current;
	}

	_destinyScript = code;
	return _scriptPtr = _destinyScript.data();
}
```

Reply:
Yes, this ``_scriptPtr`` assignment is legal here because ``_destinyScript`` is a member variable.

Reply:
Nice! I'll push again to see whether the code passes.